[Clang][BPF] Add __BPF_CPU_VERSION__ macro

yonghong-song opened the pull request:
Sometimes bpf developer might want to develop different codes
based on particular cpu versioins. For example, cpu v1/v2/v3
branch target is 16bit while cpu v4 branch target is 32bit,
thus cpu v4 allows more aggressive loop unrolling than cpu v1/v2/v3
(see [1] for a kernel selftest failure due to this).
We would like to maintain aggressive loop unrolling for cpu v4
while limit loop unrolling for earlier cpu versions.
Another example, signed divide also only available with cpu v4.

Actually, adding cpu specific macros are fairly common
in llvm. For example, x86 has maco like 'i486', '__pentium_mmx__', etc.
AArch64 has '__ARM_NEON', '__ARM_FEATURE_SVE', etc.

This patch added __BPF_CPU_VERSION__ macro. Current possible values
are 0/1/2/3/4. The following are the -mcpu=... to __BPF_CPU_VERSION__
mapping:
```
       cpu                  __BPF_CPU_VERSION__
       no -mcpu=<...>       1
       -mcpu=v1             1
       -mcpu=v2             2
       -mcpu=v3             3
       -mcpu=v4             4
       -mcpu=generic        1
       -mcpu=probe          0
```
    
This patch also added some macros for developers to identify some cpu
insn features:
```
      feature macro               enabled in which cpu
      __BPF_FEATURE_JMP_EXT       >= v2
      __BPF_FEATURE_JMP32         >= v3
      __BPF_FEATURE_ALU32         >= v3
      __BPF_FEATURE_LDSX          >= v4
      __BPF_FEATURE_MOVSX         >= v4
      __BPF_FEATURE_BSWAP         >= v4
      __BPF_FEATURE_SDIV_SMOD     >= v4
      __BPF_FEATURE_GOTOL         >= v4
      __BPF_FEATURE_ST            >= v4
```    
         [1] https://lore.kernel.org/bpf/3e3a8a30-dde0-43a1-981e-2274962780ef@linux.dev/

yonghong-song commented:
cc @4ast @anakryiko

eddyz87 reviewed (approved):
Hi Yonghong,

Thank you for working on this.

4ast commented on clang/lib/Basic/Targets/BPF.cpp:
```diff
@@ -29,6 +29,14 @@ void BPFTargetInfo::getTargetDefines(const LangOptions &Opts,
+  else if (CPU.empty() || CPU == "generic")
+    Builder.defineMacro("__bpf_cpu_version__", "1");
+  else
+    Builder.defineMacro("__bpf_cpu_version__", CPU.substr(1));
```
That works and somewhat similar to other archs. Like __amdgcn_processor__.
I have slight preference to use capitol case __BPF_CPU_VERSION__.
Let's add all groups too:
    HasJmpExt = true;
    HasJmp32 = true;
    HasAlu32 = true;
    HasJmpExt = true;
    HasJmp32 = true;
    HasAlu32 = true;
    HasLdsx = !Disable_ldsx;
    HasMovsx = !Disable_movsx;
    HasBswap = !Disable_bswap;
    HasSdivSmod = !Disable_sdiv_smod;
    HasGotol = !Disable_gotol;

arm does __ARM_FEATURE_xx.
We can do __BPF_FEATURE_ALU32, __BPF_FEATURE_GOTOL.

yonghong-song replied:
Some architectures prefer small case (x86, pcc, loongarch, riscv) while some other architectures (arm, etc.) prefer upper case. Yes, we can use capitol cases for all the proposed macros.
For BPF_CPU_VERSION, I would like __BPF_CPU_VERSION. Basically adding '__' as prefix. What do you think?

yonghong-song replied:
Okay, during coding, I feel that I don't really like __BPF_CPU_VERSION. I would prefer either BPF_CPU_VERSION or __BPF_CPU_VERSION__. I will tentatively use BPF_CPU_VERSION for now.

yonghong-song replied:
Okay, I don't particularly like __BPF_CPU_VERSION either and actually prefer double underscore before and after BPF_CPU_VERSION. Now, I realize that you actually mean with double underscore before and after BPF_CPU_VERSION. github just displays as highlight.

Commit:
[Clang][BPF] Add __BPF_CPU_VERSION__ macro

Sometimes bpf developer might want to develop different codes
based on particular cpu versioins. For example, cpu v1/v2/v3
branch target is 16bit while cpu v4 branch target is 32bit,
thus cpu v4 allows more aggressive loop unrolling than cpu v1/v2/v3
(see [1] for a kernel selftest failure due to this).
We would like to maintain aggressive loop unrolling for cpu v4
while limit loop unrolling for earlier cpu versions.
Another example, signed divide also only available with cpu v4.

Actually, adding cpu specific macros are fairly common
in llvm. For example, x86 has maco like 'i486', '__pentium_mmx__', etc.
AArch64 has '__ARM_NEON', '__ARM_FEATURE_SVE', etc.

This patch added __BPF_CPU_VERSION__ macro. Current possible values
are 0/1/2/3/4. The following are the -mcpu=... to __BPF_CPU_VERSION__
mapping:
   cpu                  __BPF_CPU_VERSION__
   no -mcpu=<...>       1
   -mcpu=v1             1
   -mcpu=v2             2
   -mcpu=v3             3
   -mcpu=v4             4
   -mcpu=generic        1
   -mcpu=probe          0

This patch also added some macros for developers to identify some cpu
insn features:
  feature macro               enabled in which cpu
  __BPF_FEATURE_JMP_EXT       >= v2
  __BPF_FEATURE_JMP32         >= v3
  __BPF_FEATURE_ALU32         >= v3
  __BPF_FEATURE_LDSX          >= v4
  __BPF_FEATURE_MOVSX         >= v4
  __BPF_FEATURE_BSWAP         >= v4
  __BPF_FEATURE_SDIV_SMOD     >= v4
  __BPF_FEATURE_GOTOL         >= v4
  __BPF_FEATURE_ST            >= v4

     [1] https://lore.kernel.org/bpf/3e3a8a30-dde0-43a1-981e-2274962780ef@linux.dev/
